Sources/PackageRegistry/RegistryClient.swift

```diff
@@ -655,50 +705,160 @@ public final class RegistryClient: Cancellable {
-                                switch $0 {
-                                case .archive:
-                                    return .archive
-                                case .request:
-                                    return .request
+                        return PublishRequirements(
+                            metadata: .init(
+                                location: publishRequirements.metadata.location.map {
+                                    switch $0 {
+                                    case .archive:
+                                        return .archive
+                                    case .request:
+                                        return .request
+                                    }
                                 }
-                            }
-                        ),
-                        signing: .init(
-                            required: publishRequirements.signing.required,
-                            acceptedSignatureFormats: publishRequirements.signing.acceptedSignatureFormats.map {
-                                switch $0 {
-                                case .CMS_1_0_0:
-                                    return .CMS_1_0_0
-                                }
-                            },
-                            trustedRootCertificates: publishRequirements.signing.trustedRootCertificates
+                            ),
+                            signing: .init(
+                                required: publishRequirements.signing.required,
+                                acceptedSignatureFormats: publishRequirements.signing.acceptedSignatureFormats.map {
+                                    switch $0 {
+                                    case .CMS_1_0_0:
+                                        return .CMS_1_0_0
+                                    }
+                                },
+                                trustedRootCertificates: publishRequirements.signing.trustedRootCertificates
+                            )
                         )
-                    )
+                    default:
+                        if let error = try? response.parseError(decoder: self.jsonDecoder) {
+                            throw RegistryError.serverError(code: response.statusCode, details: error.detail)
+                        } else {
+                            throw RegistryError.invalidResponseStatus(expected: [200], actual: response.statusCode)
+                        }
+                    }
                 }.mapError {
                     RegistryError.failedRetrievingRegistryPublishRequirements($0)
                 }
             )
         }
     }
 
+    public func publish(
+            registryURL: URL,
+            packageIdentity: PackageIdentity,
+            packageVersion: Version,
+            packageArchive: AbsolutePath,
+            packageMetadata: AbsolutePath?,
```

Comment:
```suggestion
            packageMetadata: AbsolutePath?,
            signature: Data?
```

I also wonder if `packageArchive` and `packageMetadata` should be `Data` instead of path here.

Reply:
my plan with the path here is that we can avoid loading the file into memory and instead stream it into the upload request. was planning to do this in a follow up PR